src/data/common/Dhis2DataFormRepository.ts

```diff
@@ -164,6 +166,80 @@ export class Dhis2DataFormRepository implements DataFormRepository {
         });
     }
 
+    private getTotalRules(sections: Section[], dataElements: DataElement[]): TotalRule[] {
+        return sections.flatMap(section => {
+            const totalsFormulas = _(section.totals?.formulas).values();
+
+            return totalsFormulas
+                .map(formulaRules => {
+                    const { rules, formula } = formulaRules;
+                    if (!rules) return undefined;
+
+                    const visibleTotalRule = this.getTotalRuleByRuleType(
+                        "visible",
+                        rules,
+                        section,
+                        dataElements,
+                        formula
+                    );
+                    const disabledTotalRule = this.getTotalRuleByRuleType(
+                        "disabled",
+                        rules,
+                        section,
+                        dataElements,
+                        formula
+                    );
+
+                    return visibleTotalRule || disabledTotalRule;
+                })
+                .compact()
+                .value();
+        });
+    }
+
+    private getTotalRuleByRuleType(
+        ruleType: RuleType,
+        rules: DataElementRuleOptions,
+        section: Section,
+        dataElements: DataElement[],
+        formula: string
+    ): Maybe<TotalRule> {
+        const relatedDataElements = this.getRelatedDataElements(section, dataElements, formula);
+        const rule = rules[ruleType];
+        if (!rule) return undefined;
+
+        return {
+            type: ruleType,
+            dataElements: rule.dataElements.map(
+                dataElementCode => dataElements.find(de => de.code === dataElementCode)?.id ?? ""
+            ),
+            relatedDataElements: relatedDataElements,
+            condition: rule.condition,
+            formula: formula,
+        };
+    }
+
+    private getRelatedDataElements(section: Section, dataElements: DataElement[], formula: string): DataElement[] {
+        const dataElementCodesInFormula = formula.match(/\b[A-Za-z_][A-Za-z0-9_]*\b/g) || [];
+        const dataElementsByCode = _(dataElements)
```

Comment:
Sorry I wasn't clear enough on this. This is correct, but can we do it outside the loop to avoid re-generating the dataElementsByCode for every section?

```ts
private getTotalRules(sections: Section[], dataElements: DataElement[]): TotalRule[] {
    const dataElementsByCode = _(dataElements)
          .keyBy(de => de.code)
          .value();
          
          // then pass Record<Code, DataElement> instead DataElement[] into the methods
}

```

Reply:
oh i understand now. thanks Eduardo! :)